packages/fuels-signers/src/predicate.rs

```diff
@@ -0,0 +1,131 @@
+use std::{fmt::Debug, fs};
+
+use fuel_tx::Contract;
+use fuel_types::{Address, AssetId};
+use fuels_types::{
+    bech32::Bech32Address, constants::BASE_ASSET_ID, errors::Result, input::Input,
+    transaction_builders::TransactionBuilder, unresolved_bytes::UnresolvedBytes,
+};
+
+use crate::{
+    accounts_utils::{adjust_inputs, adjust_outputs, calculate_base_amount_with_fee},
+    provider::Provider,
+    Account, AccountError, AccountResult,
+};
+
+#[derive(Debug, Clone)]
+pub struct Predicate {
+    address: Bech32Address,
+    code: Vec<u8>,
+    pub data: UnresolvedBytes,
+    pub provider: Option<Provider>,
+}
+
+impl Predicate {
+    pub fn address(&self) -> &Bech32Address {
+        &self.address
+    }
+
+    pub fn set_predicate(&mut self, provider: Provider) -> &mut Self {
```

Comment:
```suggestion
    pub fn set_provider(&mut self, provider: Provider) -> &mut Self {
```

Reply:
I took care of this not only here but also in other locations.

Reply:
@hal3e  :laughing: so much talk about predicates went to my head :D